notebook2/notebook/notebook.py

```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
+    """
+        Summarize the container status based on its most recent state
+
+        Parameters
+        ----------
+        container_statuses : list[V1ContainerStatus]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md
+    """
+    if container_statuses is None:
+        return None
+
+    state = container_statuses[0].state
+    rn = None
+    wt = None
+    tm = None
+    if state.running:
+        rn = {"started_at": state.running.started_at}
+
+    if state.waiting:
+        wt = {"reason": state.waiting.reason}
+
+    if state.terminated:
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
+    """
+        Return the most recent status=="True" V1PodCondition or None
+
+        Parameters
+        ----------
+        conds : list[V1PodCondition]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodCondition.md
+
+    """
+    if conds is None:
+        return None
+
+    maxDate = None
+    maxCond = None
+    for condition in conds:
+        if maxDate is None:
+            maxCond = condition
+            maxDate = condition.last_transition_time
+            continue
+
+        if condition.last_transition_time > maxDate and condition.status == "True":
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
+
+
+def marshall_notebooks(pods = [], svc_status = None):
+    notebooks = []
+
+    for pod in pods:
+        notebook = {
+            'name': pod.metadata.labels['name'],
+            'pod_name': pod.metadata.name,
+            'svc_name': pod.metadata.labels['svc_name'],
+            'pod_status': pod.status.phase,
+            'creation_date': pod.metadata.creation_timestamp.strftime('%D'),
+            'jupyter_token': pod.metadata.labels['jupyter_token'],
+            'container_status': read_containers_status(pod.status.container_statuses),
+            'condition': read_conditions(pod.status.conditions)
+        }
+
+        notebook['url'] = f"/instance/{notebook['svc_name']}/?token={notebook['jupyter_token']}"
+
+        if svc_status is not None:
+            notebook['svc_status'] = svc_status
+        else:
+            notebook['svc_status'] = read_svc_status(notebook['svc_name'])
+
+        notebooks.append(notebook)
+
+    return notebooks
+
+
+def get_live_user_notebooks(user_id = None):
```

Comment:
I generally prefer to return null/nil/None when a resource doesn't exist

Reply:
Using `None` instead of `[]` adds unnecessary condition checks when we're doing loop comprehensions.

Using both `None` and `[]` means we have two redundant ways to represent emptiness.

I don't see a compelling reason for using `None` instead of `[]` or both unless one of them has a distinct meaning.